palace/linalg/errorestimator.hpp

```diff
@@ -0,0 +1,91 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PALACE_LINALG_ERROR_ESTIMATOR_HPP
+#define PALACE_LINALG_ERROR_ESTIMATOR_HPP
+
+#include <mfem.hpp>
+
+#include "linalg/fluxprojector.hpp"
+#include "utils/errorindicators.hpp"
+
+namespace palace
+{
+
+class IoData;
+class MaterialOperator;
+
+namespace petsc
+{
+class PetscParVector;
+}  // namespace petsc
+
+// Class used for computing curl flux error estimate,
+// i.e. || μ⁻¹∇ × Vₕ - F ||_K
+// where F denotes a smooth reconstruction of μ⁻¹∇ × Vₕ.
+class CurlFluxErrorEstimator
+{
+  const MaterialOperator &mat_op;
+  // The finite element space used to represent V.
+  mfem::ParFiniteElementSpace &fes;
+
+  std::vector<std::unique_ptr<mfem::ND_FECollection>> smooth_flux_fecs;
+  mutable mfem::ParFiniteElementSpaceHierarchy smooth_flux_fes;
+  mutable FluxProjector smooth_projector;
+
+  mfem::L2_FECollection coarse_flux_fec;
+  mutable mfem::ParFiniteElementSpace coarse_flux_fes;
+
+  std::vector<mfem::DenseMatrix> scalar_mass_matrices;
+  std::vector<mfem::DenseMatrix> smooth_to_coarse_embed;
+
+public:
+  // Constructor for using geometric and p multigrid.
+  CurlFluxErrorEstimator(const IoData &iodata, const MaterialOperator &mat_op,
+                         const std::vector<std::unique_ptr<mfem::ParMesh>> &mesh,
+                         mfem::ParFiniteElementSpace &fes);
+
+  // Compute elemental error indicators given a complex vector of true DOF.
+  IndicatorsAndNormalization operator()(const ComplexVector &v,
```

Comment:
I don't see the downside of giving these functions names so when you call them it is more clear what's going on.